### ISSUE TITLE
--trim-filenames gets confused by directory name in --output

### ISSUE THREAD

**Issue by axcore** [yt-dlp/yt-dlp#3494]
### Checklist

- [X] I'm reporting a bug unrelated to a specific site
- [X] I've verified that I'm running yt-dlp version **2022.04.08** ([update instructions](https://github.com/yt-dlp/yt-dlp#update)) or later (specify commit)
- [X] I've checked that all provided URLs are alive and playable in a browser
- [X] I've checked that all URLs and arguments with special characters are [properly quoted or escaped](https://github.com/ytdl-org/youtube-dl#video-url-contains-an-ampersand-and-im-getting-some-strange-output-1-2839-or-v-is-not-recognized-as-an-internal-or-external-command)
- [X] I've searched the [bugtracker](https://github.com/yt-dlp/yt-dlp/issues?q=) for similar issues including closed ones. DO NOT post duplicates
- [X] I've read the [guidelines for opening an issue](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#opening-an-issue)

### Description

`yt-dlp --trim-filenames 5 https://www.youtube.com/watch?v=jNQXAC9IVRw`

This produces the expected file, whose filename is 5 characters long: `Me at.webm`

`yt-dlp --trim-filenames 5 --output "testfolder/%(title)s.%(ext)s" https://www.youtube.com/watch?v=jNQXAC9IVRw`

This should produce the file `../testfolder/Me at.webm`, but instead it produces the file `testf.webm`.

Worse still, if the `--output` argument is an absolute path like /home/myname/folder/folder/stuff, yt-dlp will try to create a file called /home/.webm (in the root directory).

So, I think `--trim-filenames` is broken. But if trimming the directory name is actually the intended behaviour, then it should probably be renamed to `--trim-paths` or something like that.

### Verbose log

```shell
yt-dlp --trim-filenames 5 --output "testfolder/%(title)s.%(ext)s" https://www.youtube.com/watch?v=jNQXAC9IVRw
[youtube] jNQXAC9IVRw: Downloading webpage
[youtube] jNQXAC9IVRw: Downloading android player API JSON
[info] jNQXAC9IVRw: Downloading 1 format(s): 242+251
[download] Destination: testf.f242.webm
[download] 100% of 378.64KiB in 00:00
[download] Destination: testf.f251.webm
[download] 100% of 225.23KiB in 00:00
[Merger] Merging formats into "testf.webm"
Deleting original file testf.f242.webm (pass -k to keep)
Deleting original file testf.f251.webm (pass -k to keep)
```

**Comment by pukkandan** [comment]
I agree that this is not intuitive. But this behavior is intentional since the option was added to counter the path length issues in Windows https://github.com/blackjack4494/youtube-dlc/issues/85

I honestly recommend using something like `-o %(title).200B.%(ext)s` instead to control the filename length

**Comment by raceviper13a** [comment]
This should be fixed.  At least the documentation should be updated to describe what's actually happening and demonstrate how to get what the user wants.

**Comment by chapmanjacobd** [comment]
I think to "fix" this you would just need to operate on the filename and not the path:

    path = filename  # for clarity in this comment
    basename, filename = os.path.split(path)

    no_ext, *ext = filename.rsplit('.', 2)
    filename = join_nonempty(no_ext[:trim_file_name], *ext, delim='.')

    path = os.path.join(basename, filename)

I agree the current behavior of `--trim-filenames` is doing something weird if `outtmpl` contains paths:

'path/to/test.txt' becomes 'pa.txt' instead of 'path/to/te.txt' with `--trim-filenames 2`

I imagine the current implementation might be due to a misunderstanding of the Windows limitations. It looks like with NTFS and ReFS the limit is mostly on the file name. Not the directory fragments:

- https://en.wikipedia.org/wiki/Long_filename#Limits
- 32767 / 255 = ~128 nested folders with long names...

**Comment by Grub4K** [comment]
It is non trivial to operate on the filename only. We have lost the metadata at that point, and cannot decide if something is an extension easily. Given the input of `Daily news (11.05.2024) [ID].mp4` for example, your code would treat the `.2024) [ID]` as an extension. Also the `id` portion should ideally never be touched to allow for identification of that file later on.

Additionally, you have problems where for example people would want to use individual path segments like `%(series)s S%(season_number)d - %(season_name)s` which *could* grow larger. Without more information, its hard to tell where we should trim.

Behavior would have to properly handle these cases, and path trimming would have to be done *during* output template substitution so we can know exactly what we could trim and what we couldn't, based on heuristics.

**Comment by chapmanjacobd** [comment]
> Also the id portion should ideally never be touched to allow for identification of that file later on.

I agree. The above code two lines are copied from master. But I think that portion of the problem is fixed in https://github.com/yt-dlp/yt-dlp/pull/7954. This is about path separators in outtmpl and documentation of --trim-filenames functionality

**Comment by hunter-gatherer8** [comment]
Maybe it should be a separate issue, but since this seems to be a generic issue about why `--trim-filenames` doesn't make sense, I'll report it here.

The use case is `ecryptfs`, which has a limitation of 255 bytes per filename (filepath length doesn't matter). `--trim-filenames` looks like it exists just for that, but it doesn't handle it for multiple reasons:
- Does not account for `%(ext)s`: that's the least important problem, since one can just set the limit some 17 characters less (and it's clearly stated in the docs).
- As mentioned above, it counts total `-o` length, not filename, which doesn't make any sense regardless of FS -- it's **neither** the filename length, nor total filepath length, unless `-o` is absolute.
- Worst of all, it counts **characters, not bytes**, so it still won't fit within the defined limit if it has any non-ASCII chars.

> I honestly recommend using something like `-o %(title).200B.%(ext)s` instead to control the filename length

@pukkandan this only solves the problem if you have nothing but `title` in the filename, so this really isn't the same use case.
